deeplinkdispatch-processor/src/main/java/com/airbnb/deeplinkdispatch/UrlTrie.kt

```diff
@@ -0,0 +1,143 @@
+    }
+
+    /**
+     * Byte array format is:
+     * 0                                                    type
+     * 1                                                    value length
+     * 2..5                                                children length
+     * 6..8                                                 match id
+     * 8..(8+value length)                                  value
+     * (8+value length)..((8+value length)+children length) children
+     */
+    fun toByteArray(): UByteArray {
+        val childrenByteArrays: List<UByteArray> = generateChildrenByteArrays()
+        val valueByteArray = (if (placeholder) IDX_PLACEHOLDER + id else id ).let { it.toByteArray(UTF_8).toUByteArray() }
+        val header = generateHeader(type, valueByteArray, childrenByteArrays, match)
+        val resultByteArray = UByteArray(arrayLength(
+                childrenByteArrays,
+                valueByteArray,
+                header
+        ))
+        header.copyInto(resultByteArray)
+        var position = header.size
+        with(valueByteArray){
+            copyInto(resultByteArray, position)
+            position += size
+        }
+        for (childByteArray in childrenByteArrays) {
+            childByteArray.copyInto(resultByteArray, position)
+            position += childByteArray.size
+        }
+        return resultByteArray
+    }
+
+    private fun arrayLength(childArrays: List<UByteArray>, value: UByteArray, header: UByteArray): Int {
+        var length = header.size + value.size
+        childArrays.forEach { child ->
+            length += child.size
+        }
+        return length
+    }
+
+    private fun generateChildrenByteArrays(): List<UByteArray> = children.map { it.toByteArray() }
+
+    private fun generateHeader(type: UByte, value: UByteArray, children: List<UByteArray>? = null, match: UriMatch?): UByteArray {
+        var childrenLength: UInt = 0u
+        children?.forEach { child ->
+            childrenLength += child.size.toUInt()
+        }
+        val header = UByteArray(HEADER_LENGTH)
+        header.set(0, type)
+        header.set(1, value.size.toUByte()) // If this is a placeholder we will only save one byte
+        header.writeUIntAt(2, childrenLength)
+        header.writeUShortAt(6, if (match == null) UShort.MAX_VALUE else match.matchId.toUShort())
+        return header
+    }
+
+}
+
+private val MAX_EXPOT_STRING_SIZE = 60000
+
+data class Root(val bla: String = "") : TrieNode(ROOT_VALUE, TYPE_ROOT.toUByte()) {
+    fun writeToOutoutStream(openOutputStream: OutputStream) {
+        openOutputStream.write(this.toByteArray().toByteArray())
+    }
+
+    /**
+     * Convert the byte array into a string return as max 60k length sset of strings.
+     */
+    fun getStrings(): List<String> {
+        return String(bytes = this.toByteArray().toByteArray(), charset = Charset.forName(MatchIndex.MATCH_INDEX_ENCODING)).chunked(MAX_EXPOT_STRING_SIZE)
```

Comment:
`toByteArray().toByteArray()` duplicate?

Reply:
No, the naming is a problem. I'll fix the naming